File: 113.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def pathSum(self, root, sum):
        """
        :type root: TreeNode
        :type sum: int
        :rtype: List[List[int]]
        """
        if not root: return []
        res = []
        self.dfs(root, sum, [], res)
        return res
    def dfs(self,root, sum, path, res):
        if not root:
            return
        if not root.left and not root.right and root.val == sum:
            path.append(root.val)
            res.append(path)
            return
        if root.left:
            self.dfs(root.left, sum-root.val, path+[root.val], res)
        if root.right:
            self.dfs(root.right, sum-root.val, path+[root.val], res)

File: 113.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> pathSum(TreeNode* root, int targetSum) {
        vector<vector<int>> res;
        vector<int> tmp;
        dfs(root, res, tmp, targetSum);
        return res;
    }
    void dfs(TreeNode* root, vector<vector<int>>& res, vector<int>& tmp, int target){
        if (root == NULL){
            return;
        }
        if (root->left == NULL and root->right ==NULL and root->val == target){
            tmp.push_back(root->val);
            res.push_back(tmp);
            tmp.pop_back();
            return;
        }
        tmp.push_back(root->val);
        dfs(root->left, res, tmp, target-root->val);
        dfs(root->right, res, tmp, target - root->val);
        tmp.pop_back();
    }
};
